### PR TITLE
Add compatibility logic for jsonschema

### PR DESCRIPTION
**Pull request by gabriel-tincu**
- Add base for backward / forward compatibility check for base and complex types
- Add more logging and basic validation rules for forwarding base types
- Validation for primitive types and most cases i could think of on the spot for complex types
- Unit tests for compatibility
- Extract check calls into separate functions, tested against confluent rest
- Update consumer tests to also test jsonschema serialization
- Default kafka port for fixtures running with pre existing kafka service

NOTE: This PR is open against a non master branch for diff viewing purposes. IT will be reopen against master when the changes on the base branch are added to it, and left as a draft until then

### CODE REVIEW

File: karapace/compatibility.py [hunk 1]
```diff
@@ -39,12 +46,226 @@ def check(self) -> bool:  # pylint: disable=inconsistent-return-statements
...
-        raise NotImplementedError("write me")
+    @staticmethod
+    def are_same_type(source: Union[dict, str], target: Union[dict, str]) -> bool:
+        return JsonSchemaCompatibility.get_type_info(source) == JsonSchemaCompatibility.get_type_info(target)
+
+    @staticmethod
+    def get_type_info(source: Union[dict, str]) -> (bool, str):
+        """
+        :param source: a type object, either a string for primitives or a dict for all
+        :return: tuple representing weather the type is complex or base on the first element and the type name on the second
```
Comment (Excds) [inline karapace/compatibility.py:63]: Typo.

```suggestion
        :return: tuple representing whether the type is complex or base on the first element and the type name on the second
```

File: karapace/compatibility.py [hunk 2]
```diff
@@ -39,12 +46,226 @@ def check(self) -> bool:  # pylint: disable=inconsistent-return-statements
...
 
 class JsonSchemaCompatibility(Compatibility):
-    def check_compatibility(self, source: TypedSchema, target: TypedSchema) -> bool:
-        raise NotImplementedError("write me")
+    @staticmethod
+    def are_same_type(source: Union[dict, str], target: Union[dict, str]) -> bool:
+        return JsonSchemaCompatibility.get_type_info(source) == JsonSchemaCompatibility.get_type_info(target)
+
+    @staticmethod
+    def get_type_info(source: Union[dict, str]) -> (bool, str):
```
Comment (Excds) [inline karapace/compatibility.py:60]: Can you add a parameterized test for this method?

File: karapace/compatibility.py [hunk 3]
```diff
@@ -39,12 +46,226 @@ def check(self) -> bool:  # pylint: disable=inconsistent-return-statements
...
+        if field_type in {"number", "integer"}:
+            return JsonSchemaCompatibility.can_read_number(source, target)
+        if field_type == "string":
+            return JsonSchemaCompatibility.can_read_string(source, target)
+        if field_type in {"boolean", "null"}:
+            return True
+        raise UnknownSchemaType(f"Unknown field type {field_type}")
+
+    @staticmethod
+    def can_read_number(source: dict, target: dict) -> bool:
```
Comment (Excds) [inline karapace/compatibility.py:93]: It would be nice if this also had a test.

File: karapace/compatibility.py [hunk 4]
```diff
@@ -39,12 +46,226 @@ def check(self) -> bool:  # pylint: disable=inconsistent-return-statements
...
+                )
+        source_min_len, target_min_len = source.get("minLength", 0), target.get("minLength", 0)
+        if source_min_len > target_min_len:
+            raise IncompatibleSchema(f"Source min len {source_min_len} is not greater than target min len {target_min_len}")
+        source_max_len, target_max_len = source.get("maxLength", 0), target.get("maxLength", 0)
+        if source_max_len < target_max_len:
+            raise IncompatibleSchema(f"Source max len {source_max_len} is not greater than target max len {target_max_len}")
+        return True
+
+    def can_read_objects(self, source: dict, target: dict, checking_for: str) -> bool:
```
Comment (Excds) [inline karapace/compatibility.py:147]: Same for this.

File: karapace/compatibility.py [hunk 5]
```diff
@@ -39,12 +46,226 @@ def check(self) -> bool:  # pylint: disable=inconsistent-return-statements
...
+        if target_additional_props is True and source_additional_props is not True:
+            raise IncompatibleSchema(f"{source} does not cover {target} on additionalProperties")
+        if isinstance(source_additional_props, dict) and isinstance(target_additional_props, dict):
+            self.log.debug("Recursing for additional properties on %r and %r", source, target)
+            self.check_compatibility(source_additional_props, target_additional_props, checking_for)
+        return True
+        # no patternProperties, no dependencies
+
+    @staticmethod
+    def can_read_unions(source, target, checking_for: str):  # pylint: disable=unused-argument
```
Comment (Excds) [inline karapace/compatibility.py:167]: Is this supposed to be only returning True?

File: karapace/compatibility.py [hunk 6]
```diff
@@ -39,12 +46,226 @@ def check(self) -> bool:  # pylint: disable=inconsistent-return-statements
...
+        return True
+
+    @staticmethod
+    def can_read_array(source: dict, target: dict, checking_for: str) -> bool:  # pylint: disable=unused-argument
```
Comment (Excds) [inline karapace/compatibility.py:171]: A test would be nice.

File: karapace/compatibility.py [hunk 7]
```diff
@@ -39,12 +46,226 @@ def check(self) -> bool:  # pylint: disable=inconsistent-return-statements
...
+        if source.get("minItems", 0) > target.get("minItems", 0):
+            raise IncompatibleSchema(f"{source} does not cover {target} on minItems")
+        if source.get("maxItems", 0) < target.get("maxItems", 0):
+            raise IncompatibleSchema(f"{source} does not cover {target} on maxItems")
+        if source.get("uniqueItems") and not target.get("uniqueItems"):
+            raise IncompatibleSchema(f"{source} has unique items enabled while {target} does not")
+        return True
+
+    def can_read_complex(self, source, target, field_type, checking_for: str) -> bool:
```
Comment (Excds) [inline karapace/compatibility.py:180]: A test would be nice.

File: karapace/compatibility.py [hunk 8]
```diff
@@ -39,12 +46,226 @@ def check(self) -> bool:  # pylint: disable=inconsistent-return-statements
...
+            can_read = validators[field_type](source, target, recurse_check_for)
+        if checking_for in {"FULL", "FORWARD"}:
+            recurse_check_for = "FULL" if checking_for == "FULL" else "BACKWARD"
+            can_be_read = validators[field_type](target, source, recurse_check_for)
+        if not can_read and can_be_read:
+            raise IncompatibleSchema(f"Fields {source} and {target} contain incompatible properties")
+        return True
+
+    @staticmethod
+    def get_fields(source):
```
Comment (Excds) [inline karapace/compatibility.py:196]: A test would be nice.

File: karapace/compatibility.py [hunk 9]
```diff
@@ -39,12 +46,226 @@ def check(self) -> bool:  # pylint: disable=inconsistent-return-statements
...
+            return [(field, props[field]) for field in props]
+        if source["type"] == "array":
+            # how do we handle contains fields??
+            fields = source["items"] if isinstance(source["items"], list) else [source["items"]]
+            return [(None, item) for item in fields]
+        if isinstance(source["type"], list):
+            return [(None, field) for field in source["type"]]
+        raise UnknownSchemaType(f"Unknown complex type: {source}")
+
+    def check_same_type(self, source, target, checking_for: str) -> bool:
```
Comment (Excds) [inline karapace/compatibility.py:208]: A test would be nice.

File: karapace/compatibility.py [hunk 10]
```diff
@@ -39,12 +46,226 @@ def check(self) -> bool:  # pylint: disable=inconsistent-return-statements
...
+
+        if (checking_for in {"FULL", "FORWARD"} or type_name == "object") and len(unhandled_source_indexes) > 0:
+            self.log.debug("Fields %r are missing from %r  for %s compatibility", source, target, self._checking_for)
+            raise IncompatibleSchema(f"{source} incompatible with {target} with the compatibility {self._checking_for}")
+        if (checking_for in {"FULL", "BACKWARD"} or type_name == "object") and len(unhandled_target_indexes) > 0:
+            self.log.debug("Fields %r are missing from %r  for %s compatibility", target, source, self._checking_for)
+            raise IncompatibleSchema(f"{target} incompatible with {source} with the compatibility {self._checking_for}")
+        return True
+
+    def handle_union(self, checking_for, source, target):
```
Comment (Excds) [inline karapace/compatibility.py:234]: A test would be nice.

File: karapace/compatibility.py [hunk 11]
```diff
@@ -39,12 +46,226 @@ def check(self) -> bool:  # pylint: disable=inconsistent-return-statements
...
+        unhandled_source_indexes = set(range(len(source_fields)))
+        for i, ((_, source_field), (_, target_field)) in enumerate(zip(source_fields, target_fields)):
+            if not self.are_same_type(source_field, target_field):
+                continue
+            self.check_compatibility(source_field, target_field, checking_for)
+            unhandled_source_indexes.remove(i)
+            unhandled_target_indexes.remove(i)
+        return unhandled_source_indexes, unhandled_target_indexes
+
+    def handle_array_or_object(self, checking_for, source, target):
```
Comment (Excds) [inline karapace/compatibility.py:247]: A test would be nice.

**Review by Excds — CHANGES_REQUESTED** [review]
I've added comments on the methods I think can use a test to validate that they're doing what they're supposed to do.

File: karapace/compatibility.py (hunk 5 above)
Comment (gabriel-tincu) [reply karapace/compatibility.py:167]: yes, unions don't have properties to check for (like required fields or min / max value)

**Comment by gabriel-tincu** [comment]
Closing this down as it needs to be totally rewritten

**Pull request closed without merging**